### PR TITLE
new release 0.8.2 of mlpost

### PR DESCRIPTION
**Pull request by backtracking**
should compile with 4.04

### CODE REVIEW

**Comment by samoht** [comment]
Doesn't seem to work on old OCaml versions:
```
#=== ERROR while installing mlpost.0.8.2 ======================================#
# opam-version 1.2.2
# os           linux
# command      make
# path         /home/travis/.opam/system/build/mlpost.0.8.2
# compiler     system (4.02.3)
# exit-code    2
# env-file     /home/travis/.opam/system/build/mlpost.0.8.2/mlpost-13018-f32074.env
# stdout-file  /home/travis/.opam/system/build/mlpost.0.8.2/mlpost-13018-f32074.out
# stderr-file  /home/travis/.opam/system/build/mlpost.0.8.2/mlpost-13018-f32074.err
### stdout ###
# Backtrace:
# [...]
#               - Building backend/dllmlpost_ft.cmx:
#                   - Failed to build all of these:
#                       - Building backend/dllmlpost_ft.ml:
#                           - Failed to build all of these:
#                               - Building backend/dllmlpost_ft.mly
#                               - Building backend/dllmlpost_ft.mll
#                       - Building backend/dllmlpost_ft.mlpack
#               - Building backend/dllmlpost_ft.mldylib
#               - Building backend/libmlpost_ft.clib
```

Does it need a new version of ocamlbuild?

File: packages/mlpost/mlpost.0.8.2/opam [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  ["ocamlfind" "remove" "mlpost"]
+  ["ocamlfind" "remove" "mlpost_lablgtk"] {"%{cairo:installed}%"}
+]
+depends: [
+  "ocamlfind"
+  "bitstring"
+  "cairo" {= "1.2.0"}
+  "ocamlbuild" {build}
+  "conf-autoconf" {build}
+]
```
Comment (avsm) [inline packages/mlpost/mlpost.0.8.2/opam]: If this is only supported on OCaml 4.04.0+, then adding:

```
available: [ ocaml-version >= "4.04.0" ]
```

will make sure users don't try to compiler with an older version of the compiler.

**Comment by backtracking** [comment]
Sorry for the confusion: my message "should compile with 4.04" simply
meant it was a fix so that it compile *also* with OCaml 4.04 (which was
not the case). But mlpost 0.8.2 compiles without trouble with older
version of OCaml, so you can merge this PR.

--
Jean-Christophe

**Comment by avsm** [comment]
Hi @backtracking -- it seems to fail with OCaml 4.02.3 and lower unfortunately.  This _may_ be due to the version of ocamlbuild being selected -- I'm not sure. cc @gasche 

```
- Backtrace:
-   - Failed to build the target backend/dllmlpost_ft.so
-       - Building backend/dllmlpost_ft.so:
-           - Failed to build all of these:
-               - Building backend/dllmlpost_ft.cmxa:
-                   - Failed to build all of these:
-                       - Building backend/dllmlpost_ft.cmx:
-                           - Failed to build all of these:
-                               - Building backend/dllmlpost_ft.ml
-                               - Building backend/dllmlpost_ft.mlpack
-                       - Building backend/dllmlpost_ft.mllib
-               - Building backend/dllmlpost_ft.cmx:
-                   - Failed to build all of these:
-                       - Building backend/dllmlpost_ft.ml:
-                           - Failed to build all of these:
-                               - Building backend/dllmlpost_ft.mly
-                               - Building backend/dllmlpost_ft.mll
-                       - Building backend/dllmlpost_ft.mlpack
-               - Building backend/dllmlpost_ft.mldylib
-               - Building backend/libmlpost_ft.clib
- ocamlbuild.Makefile:50: recipe for target 'all' failed
```

**Comment by backtracking** [comment]
Too bad. So I added the requirement for OCaml 4.04.

**Comment by gasche** [comment]
I tried with 4.03.0 (with ocamlbuild 0.11) and 4.02.3 (with 4.02.3 ocamlbuild), and the build worked in both cases in my environment. I cannot reproduce the CI failure.

**Comment by backtracking** [comment]
I did the same experiment yesterday with OCaml 4.02.3 on my machine and
I confirm I could compile mlpost 0.8.2 without trouble.

--
Jean-Christophe

**Comment by avsm** [comment]
It only happens when the cairo depopt is active -- did you try with that @gasche? Full build log: https://travis-ci.org/ocaml/opam-repository/jobs/209676118

**Comment by avsm** [comment]
I'll go ahead and merge this -- we can relax the compiler constraint later if someone debugs the ocamlbuild issue.

**Comment by gasche** [comment]
I think merging was fine -- better to have it available for 4.04 users than none at all.

That said, I do have `cairo` OCaml packages installed (in the package metadata it's just a dependency, not a depopt) and I am still unable to reproduce the failure. On my machine, `dllmlpost_ft.so` and `libmlpost_ft.a` build correctly under 4.02.3 -- `dllmlpost_ft.cmx` doesn't, but is not built. The build log for these targets is as follows:

```
# Target: backend/ml_mlpost_ft.c, tags: { bin_annot, c, cairo_yes, compile, concrete_yes, extension:c, file:backend/ml_mlpost_ft.c, include_freetype, quiet, traverse }
/home/gasche/.opam/4.02.3/bin/ocamlc.opt -ccopt -I/usr/include/freetype2 -c backend/ml_mlpost_ft.c
# Target: backend/ml_mlpost_ft.o, tags: {  }
mv ml_mlpost_ft.o backend/ml_mlpost_ft.o
# Target: backend/mlpost_ft, tags: { bin_annot, c, cairo_yes, concrete_yes, extension:a, file:backend/libmlpost_ft.a, ocamlmklib, quiet, traverse }
/home/gasche/.opam/4.02.3/bin/ocamlmklib -o backend/mlpost_ft backend/ml_mlpost_ft.o
# Compilation successful.
```

Could the problem be a missing dependency on `freetype` system packages? (I also find it weird that the OPAM package does not seem mention an (optional?) lablgtk dependency).

**Comment by backtracking** [comment]
thanks

Jean-Christophe

**Pull request merged**